spdx-utils/src/main/kotlin/SpdxExpression.kt

```diff
@@ -153,6 +154,22 @@ sealed class SpdxExpression {
      */
     open fun offersChoice(): Boolean = false
 
+    /**
+     * Applies a license [choice] to the given SPDX [subtreeMatcher].
+     * The [subtreeMatcher] must not contain more than two choices.
+     */
+    open fun applyChoice(subtreeMatcher: SpdxExpression, choice: SpdxExpression): SpdxExpression {
```

Comment:
I don't understand the API here, why is `choice` not applied to `this` but to `subtreeMatcher`? And why must it not contain more than two choices?

Reply:
> why is `choice` not applied to `this` but to `subtreeMatcher`?

Because the choice might only be a choice of a single subtree. E.g. (a OR b) AND c.
In my implementation the user would choose `b` for the subtree `a OR b` and the result would be `a AND c`.
Another approach would be to not use subtree matchers at all. E.g. if the choice is `b` it will be chosen no matter if the expression is `a OR b`, `a OR b OR c` or `(a OR b) AND c`.
I had the impression from our conversation and from https://github.com/oss-review-toolkit/ort/pull/2610 that this is the wanted behavior.

> And why must it not contain more than two choices?

Because I thought that `a OR b OR c` should be handled in two separate choices:
- `a OR b` choose `b`
- `b OR c` choose `b`
 Allowing three choices would end in even more combinations that need to be covered:
- `a OR b` choose `b`
- `b OR c` choose `b`
- `a OR b OR c` choose `c`

If you see a use case that this separation is required I'd agree to remove this limitation.

Reply:
TBH, I had a different understanding of how choices should work. In my view, a valid choice should always be checked against the *whole* expression. There should be no such thing as choices on sub-expressions.

That is, if the original expression would be `(a OR b) AND c`, then `b` would not be a valid choice, as it omits `c`. Valid choices would be `a AND c` or `b AND c`.

Reply:
@mnonnenmacher do you agree with @sschuberth?
I think this would make the API and the implementation easier and the configuration for users more straight forward.

Reply:
I'm a bit surprise that making choices for sub-trees is being questioned again, as we discussed this in the meeting and I believe it became clear that it should be supported.

Picking up Sebastian's example `(a OR b) AND c` I believe the choice `(a OR b) ->  b` should be applicable to it and resolve it to `b AND c´.

Reply:
I guess I'm simply confused about the API, even after looking at the tests, mostly due to variable naming I believe. I *can* see a value of being able to first "narrow down" the choice, and incrementally doing that to get a final choice which doe snot contain an `OR` anymore.

What would help at least me for a better understanding would be to change the function as follows:

    /**
     * Apply a license [choice], optionally limited to the given [subExpression], and return an [SpdxExpression] where
     * the choice is resolved.
     */
    open fun applyChoice(choice: SpdxExpression, subExpression: SpdxExpression = this): SpdxExpression {

Reply:
> > And why must it not contain more than two choices?
> 
> Because I thought that `a OR b OR c` should be handled in two separate choices:

IMO, ideally the API would work in a way that I can generically say "whenever I come across sub-expression *X*, choose *Y*", so I should be able to say "whenever I come across sub-expression `a OR b OR c`, choose `b`" or also "whenever I come across sub-expression `a OR b OR c`, choose `b OR c`".

Reply:
> IMO, ideally the API would work in a way that I can generically say "whenever I come across sub-expression _X_, choose _Y_", so I should be able to say "whenever I come across sub-expression `a OR b OR c`, choose `b`" or also "whenever I come across sub-expression `a OR b OR c`, choose `b OR c`".

Fully agree, that seem 100% inline with what I envisioned as well.

Reply:
I changed the API to your suggestions and tried to improve the readability.
Please also have a look at the tests that cover everything that I thought of. If a test doesn't exist I forgot that specific use case.

Short summary how the API will behave:
- Whenever I come across sub-expression `a OR b OR c`, choose `b`
- Whenever I come across sub-expression `a OR b OR c`, choose `b OR c`
- If the sub-expression does not match the expression, the expression is returned (should we use 
- Throws an exception if the choice is not valid for the subExpression

Reply:
@fviernau, @sschuberth please have another look at the changes